(QENG-4823) AWS EC2 Reaper

tvpartytonight opened the pull request:
This is the initial commit for the AWS EC2 Reaper.

tvpartytonight committed:
Initial commit

Commit:
(QENG-4823) AWS EC2 Reaper

This is the initial commit for the AWS EC2 Reaper.

tvpartytonight commented:
@binford2k @doug-rosser I'm going to have the QE team review this, but I just wanted to get more eyes on this at this stage.

tvpartytonight commented:
This is a mostly done work in progress; the scripts themselves have the destructive parts commented out at the moment, so running them in ec2 is essentially a no-op.

doug-rosser commented on lambdas/ec2/README.md:
```diff
@@ -0,0 +1,39 @@
+## Components
+
+### Schema Enforcer
+The Schema Enforcer is a lambda that is designed to be triggered when an EC2 instance enters the 'pending' state. This lambda waits for an ec2 instance to have a valid `lifetime` tag associated with it. When that tag is found, it converts that tag to a UTC date and adds a `termination_date` tag to the instance with that UTC date as its value. The Schema Enforcer terminates instances that do no ha ...
```
This isn't 100% clear. If a host doesn't have a "lifetime" tag, it does or does not get a "termination_date" tag?

tvpartytonight replied:
I think this description is missing a portion that is in the "Usage" section above:

>  Users can bypass this workflow with the `lifetime` key and set the `termination_date` directly

If a host doesn't have a `lifetime` tag, the schema enforcer will never process that tag and convert it into a 'termination_date' tag. However, you can supply a `termination_date` tag directly, and it will just process that `termination_date` tag without the lifetime conversion.

doug-rosser commented on lambdas/ec2/schema_enforcer.py:
```diff
@@ -0,0 +1,152 @@
+        # TODO: This code should also eventually report terminated instances
+        # to some log somewhere.
+        # ec2_instance.terminate()
+        raise Exception('EC2 instance deleted because no termination_date found within 4 minutes of creation.')
```
Sort of "smelly" that the enforcer also terminates (even though it doesn't).

tvpartytonight replied:
Yeah, the Enforcer deletes it if it doesn't comply with the rules. I think this confusion points to the fact that I should have a Rules section as well, to really spell it out.

doug-rosser commented on lambdas/ec2/schema_enforcer.py:
```diff
@@ -0,0 +1,152 @@
+    else:
+        raise ValueError("Unable to parse the unit '{0}'".format(unit))
+
+ec2 = boto3.resource('ec2')
```
What do here?

tvpartytonight replied:
This is the assignment of the boto3 abstraction for ec2; I wasn't sure if there was functionality that should exist with this outside of the lambda handler, so I left it here in the global scope of the script in case it was needed outside of the handler.

doug-rosser commented:
I'm 👍  except that a "safety" should be implemented to turn termination off/on globally.

Commit:
(QENG-4823) Add safety for destructive methods

Previous to this commit, the AWS EC2 reaper did not have an easy way to
turn on or off the destructive capabilities; this PR adds a global
switch that can be toggled by an environment variable that controls
whether or not the reaper is running with destructive capabilities.

doug-rosser commented on lambdas/ec2/schema_enforcer.py:
```diff
@@ -6,26 +6,35 @@
-    ec2_instance.load()
     if ec2_instance.tags is None:
-        return False
+        return None
```
Why the change here? It scares me.

`>>> q = None
>>> if q is False:
...   print "None is False"
... else:
...   print "None is NOT False"
... 
None is NOT False`

tvpartytonight replied:
I changed it here to reflect more accurately that the tag did not exist, not that it was a False value.

In your example, `None` is not `False`, so that is why that logic evaluates like that. Here is a more accurate snippet of `None` evaluates with `if`:

````
>>> if None:
...     print('None is True!')
... else:
...     print('None is False!')
None is False!
````

doug-rosser commented on lambdas/ec2/schema_enforcer.py:
```diff
@@ -0,0 +1,198 @@
+    Validates that an ec2 instance has a valid termination_date in the future.
+    """
+    ec2_termination_date = get_tag(ec2_instance, 'termination_date')
+    if ec2_termination_date is False:
```
See my previous comment where None does not appear to be (is) False.

tvpartytonight replied:
This is a bug here, it should be `None` instead of `False`.

doug-rosser commented:
I am officially 👍
